### PR TITLE
Add custom provider model config support

### LINKED ISSUE #833: Feature Request: Add Xiaomi (MiLM) Support and Implement Dynamic Model Loading
### Description

I am currently using the `tradingagents` framework and would like to request the inclusion of **Xiaomi's Large Language Models (MiLM)**. Additionally, to make the project more future-proof, I suggest implementing a mechanism that allows users to independently add new models without modifying the core `model_catalog.py` file.

### Why Xiaomi (MiLM)?

Xiaomi has released several capable models (MiLM-6B, MiLM-1.3B, etc.) that are increasingly relevant for developers in the Chinese ecosystem. Integrating these would expand the diversity of the "Deep" and "Quick" selection modes, especially for users seeking highly optimized Chinese-language performance.

### Proposed Changes

1. **Add Xiaomi to the Provider List**:
In `llm_clients/factory.py`, Xiaomi should be added to the `_OPENAI_COMPATIBLE` list if they use an OpenAI-compatible API, or a dedicated client should be implemented.
2. **Update the Model Catalog**:
Add a new entry for `xiaomi` or `xiaomi-cn` in `llm_clients/model_catalog.py`.
3. **Dynamic Model Support**:
Instead of relying solely on the static `MODEL_OPTIONS` dictionary, implement a way for the system to read custom model configurations from an external `.env` file or a local `custom_models.json`.

### Examples in Current Code

Currently, the `factory.py` uses a rigid check for providers:

```python
if provider_lower in _OPENAI_COMPATIBLE:
    from .openai_client import OpenAIClient
    return OpenAIClient(model, base_url, provider=provider_lower, **kwargs)

```

And `model_catalog.py` requires manual entry for every new model:

```python
MODEL_OPTIONS: ProviderModeOptions = {
    "openai": { ... },
    "anthropic": { ... },
    # Needs a new 'xiaomi' key here
}

```

### Benefit

Providing a method for users to "self-serve" new models (e.g., via a configuration file) would reduce the maintenance burden on the authors and allow the community to test new frontier models (like Xiaomi's) immediately upon release.

---

### Advice for you

Since you are working with e-commerce operations for a liquor enterprise and developing AI-driven automation tools, you might find that Xiaomi's models are particularly useful for local deployment or integration with smart hardware in the future.

**One question for you:** Do you already have a specific API endpoint for Xiaomi's models (like through a specific cloud provider), or are you planning to run them locally via a tool like Ollama?

### PR DESCRIPTION
**Pull request by AyushKar2005**
## What this PR does

This PR adds support for loading custom OpenAI-compatible LLM providers from a local config file.

The main reason for this change is issue #833. The issue asked about adding Xiaomi MiLM support, but there is no confirmed Xiaomi API endpoint in the issue yet. So instead of hardcoding Xiaomi directly and guessing the endpoint, I added a more general way where users can add Xiaomi MiLM or any other OpenAI-compatible provider by themself using a local json config.

The config file is loaded from:

```bash
~/.tradingagents/custom_models.json
```

This way users don't always need to modify `model_catalog.py` just to test a new model/provider.

## Why this approach

Right now, adding a new provider means changing multiple core files like provider dropdown, api key mapping, model catalog and factory. That works but it makes every new model request need a code change.

With this PR, users can define their own provider locally if the provider is OpenAI-compatible.

For example, users can add Xiaomi MiLM like this:

```json
{
  "providers": [
    {
      "display_name": "Xiaomi MiLM",
      "provider_key": "xiaomi",
      "base_url": "https://example.com/v1",
      "api_key_env": "XIAOMI_API_KEY",
      "api_type": "openai_compatible",
      "models": {
        "quick": [
          ["MiLM 1.3B", "milm-1.3b"],
          ["Custom model ID", "custom"]
        ],
        "deep": [
          ["MiLM 6B", "milm-6b"],
          ["Custom model ID", "custom"]
        ]
      }
    }
  ]
}
```

The endpoint above is only an example for testing. Users should replace it with their actual OpenAI-compatible endpoint.

## Changes made

* Added new helper for reading custom providers from `~/.tradingagents/custom_models.json`
* Added validation/normalization for custom provider keys and model options
* Added custom providers to the CLI provider dropdown
* Added support for custom provider API key env vars
* Added custom models into model selection flow
* Added custom providers into known model list
* Added support for custom OpenAI-compatible providers in the LLM client factory
* Kept `"Custom model ID"` fallback so users can still type any model manually

## Testing done

I tested this manually with a local Xiaomi MiLM config.

Created local test config:

```bash
mkdir -p ~/.tradingagents

cat > ~/.tradingagents/custom_models.json <<'JSON'
{
  "providers": [
    {
      "display_name": "Xiaomi MiLM",
      "provider_key": "xiaomi",
      "base_url": "https://example.com/v1",
      "api_key_env": "XIAOMI_API_KEY",
      "api_type": "openai_compatible",
      "models": {
        "quick": [
          ["MiLM 1.3B", "milm-1.3b"],
          ["Custom model ID", "custom"]
        ],
        "deep": [
          ["MiLM 6B", "milm-6b"],
          ["Custom model ID", "custom"]
        ]
      }
    }
  ]
}
JSON
```

Then ran the CLI:

```bash
python -m cli.main
```

Verified these things:

* Xiaomi MiLM showed up in the provider selection list
* CLI asked for `XIAOMI_API_KEY`
* Quick model selector showed `MiLM 1.3B`
* Deep model selector showed `MiLM 6B`
* CLI moved ahead to the analysis screen after selecting the models

Also ran compile check:

```bash
python -m compileall cli/utils.py tradingagents/llm_clients/*.py
```

No compile errors came from these changed files.

Screenshot 2026-06-03 212337

## Notes

This PR does not hardcode an actual Xiaomi endpoint because I could not confirm one from the issue. Instead it adds the dynamic provider loading layer, so Xiaomi MiLM or similar providers can be added locally once the user has a valid OpenAI-compatible endpoint.

Also the local `custom_models.json` is not committed. It is only used by users on their own machine.

Fixes #833

### CODE REVIEW

**Comment by praxstack** [comment]
_Automated principal-engineer review (generated by Claude Code). Findings below were adversarially verified; flagging for your judgment, not as blockers._

Thanks for this — the custom-provider config layer is cleanly factored and the input validation is a nice touch. A few things worth a look before merge, grouped by severity.

## Critical

### The configured `api_key_env` is never read at runtime; auth silently falls back to `OPENAI_API_KEY`
**File:** `tradingagents/llm_clients/openai_client.py`

This is the headline capability of the PR, but it's non-functional for custom providers. The factory routes custom providers through `OpenAIClient` (`is_custom_openai_compatible_provider`) and threads `api_key_env` through `api_key_env.py`, but `openai_client.py` only resolves/injects the API key inside the `if self.provider in _PROVIDER_BASE_URL:` branch. A custom provider (e.g. `xiaomi`) isn't in `_PROVIDER_BASE_URL`, so `get_llm()` falls through to `elif self.base_url:`, which sets *only* `base_url`. The custom key is never read.

**Failure scenario:** A user configures a custom provider with `api_key_env: XIAOMI_API_KEY`, sets that env var, and runs. Authentication silently uses `OPENAI_API_KEY` instead — either failing with a confusing error or, worse, billing the wrong account.

**Suggested fix:** In `get_llm()`, when `base_url` is set and `get_api_key_env(self.provider)` is not `None`, read `os.environ[api_key_env]` into `llm_kwargs["api_key"]`, raising the same clear "env var not set" error the built-in branch uses. Add an end-to-end test through `create_llm_client`/`get_llm` asserting the configured key is honored.

## Medium

### A custom provider can silently overwrite a built-in provider's model set
**File:** `tradingagents/llm_clients/model_catalog.py`

`get_known_models()` does `known_models[provider] = ...` unconditionally, unlike `get_model_options()` and `get_api_key_env()`, which check the built-in map first. `custom_provider_config` validates `provider_key` format but doesn't exclude built-in keys, and `_llm_provider_table` dedupes against built-ins while `get_known_models` does not. Since `validators.py` consumes `get_known_models()`, a custom `openai` entry shadows the real OpenAI known-model list used by `warn_if_unknown`.

**Failure scenario:** A user adds a custom provider keyed `openai`; the real OpenAI model list is replaced, so valid OpenAI models start triggering spurious "unknown model" warnings (or invalid ones stop being caught).

**Suggested fix:** Skip custom providers whose key already exists in `MODEL_OPTIONS` (mirror the `existing_provider_keys` guard in `_llm_provider_table`), or merge rather than overwrite — ideally reject built-in-shadowing keys in `_normalize_provider`.

### No tests for the new config module or modified resolution paths
**File:** `tradingagents/llm_clients/custom_provider_config.py`

The PR adds ~166 lines plus changes to `api_key_env.py`, `model_catalog.py`, `factory.py`, and `cli/utils.py` with no automated tests, in areas the repo conventionally covers (`test_api_key_env.py`, `test_model_validation.py`, `test_ollama_base_url.py`, `test_minimax.py`). Normalization edge cases (dict vs list model items, missing `custom` fallback, malformed JSON, duplicate keys, invalid `provider_key` regex) and the api-key/base-url runtime path are untested.

**Failure scenario:** Manual CLI testing passes because the provider appears in the dropdown, but the Critical finding above ships undetected — exactly the gap unit tests would close.

**Suggested fix:** Add unit tests for `_normalize_provider`/`_normalize_models` validation, `get_custom_model_options`/`get_custom_api_key_env`, malformed-file handling, and an end-to-end factory + `get_llm` test asserting the custom `api_key_env` is honored.

## Low

### Config file is re-read and re-parsed on every accessor call
**File:** `tradingagents/llm_clients/custom_provider_config.py`

Every public accessor calls `load_custom_providers()`, which does `exists()` + `read_text` + `json.loads` each time. `is_custom_openai_compatible_provider` runs in the client factory path, so this repeats on hot paths with no memoization.

**Suggested fix:** Memoize the parsed config (e.g. `functools.lru_cache` on a private loader keyed by file mtime, invalidated on change).

### Malformed `custom_models.json` fails silently with no diagnostic
**File:** `tradingagents/llm_clients/custom_provider_config.py`

`load_custom_providers` wraps parsing in `except Exception: return []`, and `_normalize_provider`/`_normalize_models` drop invalid entries silently. A user with a JSON typo, invalid `provider_key`, or missing `base_url` just sees their provider not appear, with no explanation.

**Suggested fix:** On parse failure or dropped entries, emit a warning (logging/console) naming the file and reason, while still degrading to an empty/partial list.

## Overall

The architecture is sound and the diff is easy to follow. The Critical finding is the one true blocker — the feature's main promise (per-provider keys) doesn't work end-to-end today, and the missing tests are why it slipped through. Recommend landing the runtime key-resolution fix plus the end-to-end test, and ideally the built-in-shadowing guard, before merge; the two Low items are good follow-ups.

🤖 Automated review via Claude Code

**Comment by praxstack** [comment]
_Proposed implementation of the fixes from my earlier automated review, offered as a ready-to-apply patch. Generated by an automated principal-engineer pass; please treat as a starting point._

**What it does:** Custom provider reads its own api_key_env at runtime (no silent OPENAI_API_KEY fallback); +24 tests

The diff is **against this PR's head commit** (apply on top of the PR branch, not `main`):

```bash
git apply pr-958.patch   # from the repo root, on this PR's branch
```

<details><summary>📋 pr-958.patch (click to expand)</summary>

```diff
diff --git c/tests/test_custom_provider_config.py w/tests/test_custom_provider_config.py
new file mode 100644
index 0000000..dfaf92f
--- /dev/null
+++ w/tests/test_custom_provider_config.py
@@ -0,0 +1,334 @@
+"""Tests for user-defined OpenAI-compatible provider config.
+
+Covers normalization edge cases, the public accessors, malformed-file
+handling, the built-in-shadowing guard in get_known_models, and an
+end-to-end client path asserting the configured api_key_env is honored at
+runtime (the headline capability of the custom-provider feature).
+"""
+
+from __future__ import annotations
+
+import json
+from pathlib import Path
+
+import pytest
+
+import tradingagents.llm_clients.custom_provider_config as cpc
+
+
+@pytest.fixture
+def custom_config(tmp_path, monkeypatch):
+    """Point CUSTOM_MODELS_FILE at a writable tmp file and clear the cache.
+
+    Returns a writer callable; each write bumps the cache via clear_cache so
+    same-second rewrites are reflected.
+    """
+    cfg_path = tmp_path / "custom_models.json"
+    monkeypatch.setattr(cpc, "CUSTOM_MODELS_FILE", cfg_path)
+
+    def write(data) -> Path:
+        if isinstance(data, str):
+            cfg_path.write_text(data, encoding="utf-8")
+        else:
+            cfg_path.write_text(json.dumps(data), encoding="utf-8")
+        cpc.clear_cache()
+        return cfg_path
+
+    cpc.clear_cache()
+    yield write
+    cpc.clear_cache()
+
+
+def _provider(**overrides):
+    base = {
+        "display_name": "Xiaomi",
+        "provider_key": "xiaomi",
+        "base_url": "https://api.xiaomi.example/v1",
+        "api_key_env": "XIAOMI_API_KEY",
+        "models": {
+            "quick": [["MiMo Quick", "mimo-quick"]],
+            "deep": [["MiMo Deep", "mimo-deep"]],
+        },
+    }
+    base.update(overrides)
+    return base
+
+
+# ---- _normalize_models ----------------------------------------------------
+
+
+def test_normalize_models_accepts_list_pairs(custom_config):
+    custom_config({"providers": [_provider()]})
+    opts = cpc.get_custom_model_options("xiaomi", "quick")
+    assert ("MiMo Quick", "mimo-quick") in opts
+    # 'custom' fallback appended when not already present.
+    assert ("Custom model ID", "custom") in opts
+
+
+def test_normalize_models_accepts_dict_items(custom_config):
+    custom_config(
+        {
+            "providers": [
+                _provider(
+                    models={
+                        "quick": [{"display": "D", "value": "v1"}],
+                        "deep": [{"name": "N", "model": "v2"}],
+                    }
+                )
+            ]
+        }
+    )
+    quick = cpc.get_custom_model_options("xiaomi", "quick")
+    deep = cpc.get_custom_model_options("xiaomi", "deep")
+    assert ("D", "v1") in quick
+    assert ("N", "v2") in deep
+
+
+def test_normalize_models_empty_gets_custom_fallback(custom_config):
+    custom_config({"providers": [_provider(models={"quick": [], "deep": []})]})
+    quick = cpc.get_custom_model_options("xiaomi", "quick")
+    assert quick == [("Custom model ID", "custom")]
+
+
+def test_normalize_models_skips_invalid_items(custom_config):
+    custom_config(
+        {
+            "providers": [
+                _provider(
+                    models={
+                        "quick": [["only-one"], 42, {"display": "ok", "value": "good"}],
+                        "deep": [["D", "v"]],
+                    }
+                )
+            ]
+        }
+    )
+    quick = cpc.get_custom_model_options("xiaomi", "quick")
+    values = [v for _, v in quick]
+    assert "good" in values
+    assert "only-one" not in values
+
+
+def test_does_not_duplicate_existing_custom_option(custom_config):
+    custom_config(
+        {
+            "providers": [
+                _provider(
+                    models={
+                        "quick": [["X", "custom"]],
+                        "deep": [["D", "v"]],
+                    }
+                )
+            ]
+        }
+    )
+    quick = cpc.get_custom_model_options("xiaomi", "quick")
+    assert [v for _, v in quick].count("custom") == 1
+
+
+# ---- _normalize_provider validation --------------------------------------
+
+
+def test_missing_required_field_dropped(custom_config):
+    custom_config({"providers": [_provider(base_url=None)]})
+    assert cpc.load_custom_providers() == []
+
+
+def test_invalid_provider_key_dropped(custom_config):
+    custom_config({"providers": [_provider(provider_key="Bad Key!")]})
+    assert cpc.load_custom_providers() == []
+
+
+def test_non_openai_api_type_dropped(custom_config):
+    custom_config({"providers": [_provider(api_type="anthropic")]})
+    assert cpc.load_custom_providers() == []
+
+
+def test_missing_models_dropped(custom_config):
+    raw = _provider()
+    del raw["models"]
+    custom_config({"providers": [raw]})
+    assert cpc.load_custom_providers() == []
+
+
+def test_duplicate_keys_first_wins(custom_config):
+    p1 = _provider(display_name="First")
+    p2 = _provider(display_name="Second")
+    custom_config({"providers": [p1, p2]})
+    providers = cpc.load_custom_providers()
+    assert len(providers) == 1
+    assert providers[0]["display_name"] == "First"
+
+
+def test_provider_key_lowercased(custom_config):
+    custom_config({"providers": [_provider(provider_key="XiaoMi")]})
+    assert cpc.is_custom_openai_compatible_provider("xiaomi")
+    assert cpc.is_custom_openai_compatible_provider("XIAOMI")
+
+
+# ---- accessors ------------------------------------------------------------
+
+
+def test_get_custom_api_key_env(custom_config):
+    custom_config({"providers": [_provider()]})
+    assert cpc.get_custom_api_key_env("xiaomi") == "XIAOMI_API_KEY"
+    assert cpc.get_custom_api_key_env("nope") is None
+
+
+def test_api_key_env_optional(custom_config):
+    p = _provider()
+    del p["api_key_env"]
+    custom_config({"providers": [p]})
+    assert cpc.get_custom_api_key_env("xiaomi") is None
+    # Provider still loads (keyless gateway).
+    assert cpc.is_custom_openai_compatible_provider("xiaomi")
+
+
+def test_provider_choices_shape(custom_config):
+    custom_config({"providers": [_provider()]})
+    choices = cpc.get_custom_provider_choices()
+    assert choices == [("Xiaomi", "xiaomi", "https://api.xiaomi.example/v1")]
+
+
+# ---- malformed-file handling ---------------------------------------------
+
+
+def test_malformed_json_returns_empty_and_warns(custom_config, caplog):
+    custom_config("{ this is not valid json ")
+    with caplog.at_level("WARNING"):
+        assert cpc.load_custom_providers() == []
+    assert any("custom_models.json" in r.message or "Could not parse" in r.message
+               for r in caplog.records)
+
+
+def test_providers_not_a_list_returns_empty(custom_config):
+    custom_config({"providers": {"not": "a list"}})
+    assert cpc.load_custom_providers() == []
+
+
+def test_missing_file_returns_empty(tmp_path, monkeypatch):
+    monkeypatch.setattr(cpc, "CUSTOM_MODELS_FILE", tmp_path / "absent.json")
+    cpc.clear_cache()
+    assert cpc.load_custom_providers() == []
+
+
+def test_dropped_entry_warns(custom_config, caplog):
+    custom_config({"providers": [_provider(provider_key="Bad Key!")]})
+    with caplog.at_level("WARNING"):
+        cpc.load_custom_providers()
+    assert any("provider_key" in r.message for r in caplog.records)
+
+
+# ---- caller mutation isolation -------------------------------------------
+
+
+def test_returned_entries_are_copies(custom_config):
+    custom_config({"providers": [_provider()]})
+    first = cpc.load_custom_providers()
+    first[0]["display_name"] = "mutated"
+    second = cpc.load_custom_providers()
+    assert second[0]["display_name"] == "Xiaomi"
+
+
+# ---- model_catalog integration: built-in shadowing guard -----------------
+
+
+def test_get_known_models_does_not_shadow_builtin(custom_config):
+    """A custom provider keyed 'openai' must not replace the built-in list."""
+    from tradingagents.llm_clients.model_catalog import MODEL_OPTIONS, get_known_models
+
+    custom_config(
+        {
+            "providers": [
+                _provider(
+                    provider_key="openai",
+                    display_name="Rogue OpenAI",
+                    models={
+                        "quick": [["rogue", "rogue-model"]],
+                        "deep": [["rogue", "rogue-model"]],
+                    },
+                )
+            ]
+        }
+    )
+    known = get_known_models()
+    builtin_values = {
+        v for opts in MODEL_OPTIONS["openai"].values() for _, v in opts
+    }
+    assert builtin_values.issubset(set(known["openai"]))
+    assert "rogue-model" not in known["openai"]
+
+
+def test_get_known_models_includes_new_custom_provider(custom_config):
+    from tradingagents.llm_clients.model_catalog import get_known_models
+
+    custom_config({"providers": [_provider()]})
+    known = get_known_models()
+    assert "xiaomi" in known
+    assert "mimo-quick" in known["xiaomi"]
+    assert "mimo-deep" in known["xiaomi"]
+
+
+# ---- end-to-end runtime key resolution (the Critical finding) ------------
+
+
+def test_custom_provider_api_key_honored_end_to_end(custom_config, monkeypatch):
+    """A custom provider's configured api_key_env must be read at runtime.
+
+    Routes through create_llm_client -> OpenAIClient.get_llm() and asserts
+    the resolved ChatOpenAI uses the custom key and base_url, not the
+    OPENAI_API_KEY fallback.
+    """
+    from tradingagents.llm_clients import create_llm_client
+
+    custom_config({"providers": [_provider()]})
+    monkeypatch.setenv("XIAOMI_API_KEY", "sk-xiaomi-secret")
+    # Ensure no accidental OpenAI fallback masks a regression.
+    monkeypatch.setenv("OPENAI_API_KEY", "sk-openai-WRONG")
+
+    client = create_llm_client(
+        provider="xiaomi",
+        model="mimo-deep",
+        base_url="https://api.xiaomi.example/v1",
+    )
+    llm = client.get_llm()
+
+    key = llm.openai_api_key
+    key_value = key.get_secret_value() if hasattr(key, "get_secret_value") else key
+    assert key_value == "sk-xiaomi-secret"
+    assert "xiaomi" in str(llm.openai_api_base)
+
+
+def test_custom_provider_missing_key_raises(custom_config, monkeypatch):
+    """Configured api_key_env that is unset must raise a clear error."""
+    from tradingagents.llm_clients import create_llm_client
+
+    custom_config({"providers": [_provider()]})
+    monkeypatch.delenv("XIAOMI_API_KEY", raising=False)
+
+    client = create_llm_client(
+        provider="xiaomi",
+        model="mimo-deep",
+        base_url="https://api.xiaomi.example/v1",
+    )
+    with pytest.raises(ValueError, match="XIAOMI_API_KEY"):
+        client.get_llm()
+
+
+def test_custom_provider_without_key_uses_placeholder(custom_config, monkeypatch):
+    """A keyless custom gateway falls back to the placeholder key."""
+    from tradingagents.llm_clients import create_llm_client
+
+    p = _provider()
+    del p["api_key_env"]
+    custom_config({"providers": [p]})
+
+    client = create_llm_client(
+        provider="xiaomi",
+        model="mimo-deep",
+        base_url="https://api.xiaomi.example/v1",
+    )
+    llm = client.get_llm()
+    key = llm.openai_api_key
+    key_value = key.get_secret_value() if hasattr(key, "get_secret_value") else key
+    assert key_value == "ollama"
diff --git c/tradingagents/llm_clients/custom_provider_config.py w/tradingagents/llm_clients/custom_provider_config.py
index 38fd4dd..925cfd7 100644
--- c/tradingagents/llm_clients/custom_provider_config.py
+++ w/tradingagents/llm_clients/custom_provider_config.py
@@ -2,13 +2,17 @@
 
 from __future__ import annotations
 
+import functools
 import json
+import logging
 import re
 from pathlib import Path
 from typing import Any, Dict, List, Optional, Tuple
 
 ModelOption = Tuple[str, str]
 
+logger = logging.getLogger(__name__)
+
 CUSTOM_MODELS_FILE = Path.home() / ".tradingagents" / "custom_models.json"
 _PROVIDER_KEY_RE = re.compile(r"^[a-z0-9][a-z0-9_-]*$")
 
@@ -54,6 +58,10 @@ def _normalize_models(models: Any) -> Optional[Dict[str, List[ModelOption]]]:
 
 def _normalize_provider(raw: Any) -> Optional[Dict[str, Any]]:
     if not isinstance(raw, dict):
+        logger.warning(
+            "Ignoring custom provider in %s: entry is not a JSON object.",
+            CUSTOM_MODELS_FILE,
+        )
         return None
 
     display_name = raw.get("display_name") or raw.get("name")
@@ -63,17 +71,42 @@ def _normalize_provider(raw: Any) -> Optional[Dict[str, Any]]:
     api_type = raw.get("api_type", "openai_compatible")
 
     if not display_name or not provider_key or not base_url:
+        logger.warning(
+            "Ignoring custom provider %r in %s: 'display_name', "
+            "'provider_key', and 'base_url' are all required.",
+            provider_key or display_name,
+            CUSTOM_MODELS_FILE,
+        )
         return None
 
     provider_key = str(provider_key).strip().lower()
     if not _PROVIDER_KEY_RE.match(provider_key):
+        logger.warning(
+            "Ignoring custom provider %r in %s: provider_key must match %s.",
+            provider_key,
+            CUSTOM_MODELS_FILE,
+            _PROVIDER_KEY_RE.pattern,
+        )
         return None
 
     if api_type != "openai_compatible":
+        logger.warning(
+            "Ignoring custom provider %r in %s: unsupported api_type %r "
+            "(only 'openai_compatible' is supported).",
+            provider_key,
+            CUSTOM_MODELS_FILE,
+            api_type,
+        )
         return None
 
     models = _normalize_models(raw.get("models"))
     if not models:
+        logger.warning(
+            "Ignoring custom provider %r in %s: 'models' must be an object "
+            "with 'quick' and/or 'deep' model lists.",
+            provider_key,
+            CUSTOM_MODELS_FILE,
+        )
         return None
 
     return {
@@ -86,18 +119,36 @@ def _normalize_provider(raw: Any) -> Optional[Dict[str, Any]]:
     }
 
 
-def load_custom_providers() -> List[Dict[str, Any]]:
-    if not CUSTOM_MODELS_FILE.exists():
-        return []
+@functools.lru_cache(maxsize=4)
+def _load_custom_providers_cached(
+    path: str, mtime: float
+) -> Tuple[Dict[str, Any], ...]:
+    """Parse and normalize the config file.
 
+    Memoized on ``(path, mtime)`` so the file is re-read and re-parsed only
+    when it changes on disk. ``is_custom_openai_compatible_provider`` runs on
+    the client-factory hot path, so the previous read-and-parse-on-every-call
+    behavior was wasteful. Returns a tuple (immutable, hashable cache value);
+    the public wrapper converts it back to a list of dicts.
+    """
     try:
-        data = json.loads(CUSTOM_MODELS_FILE.read_text(encoding="utf-8"))
-    except Exception:
-        return []
+        data = json.loads(Path(path).read_text(encoding="utf-8"))
+    except Exception as exc:
+        logger.warning(
+            "Could not parse custom provider config %s: %s. "
+            "No custom providers will be available.",
+            path,
+            exc,
+        )
+        return ()
 
-    raw_providers = data.get("providers", [])
+    raw_providers = data.get("providers", []) if isinstance(data, dict) else None
     if not isinstance(raw_providers, list):
-        return []
+        logger.warning(
+            "Custom provider config %s has no 'providers' list; ignoring.",
+            path,
+        )
+        return ()
 
     providers: List[Dict[str, Any]] = []
     seen: set[str] = set()
@@ -109,12 +160,42 @@ def load_custom_providers() -> List[Dict[str, Any]]:
 
         key = provider["provider_key"]
         if key in seen:
+            logger.warning(
+                "Ignoring duplicate custom provider %r in %s (first wins).",
+                key,
+                path,
+            )
             continue
 
         seen.add(key)
         providers.append(provider)
 
-    return providers
+    return tuple(providers)
+
+
+def load_custom_providers() -> List[Dict[str, Any]]:
+    if not CUSTOM_MODELS_FILE.exists():
+        return []
+
+    try:
+        mtime = CUSTOM_MODELS_FILE.stat().st_mtime
+    except OSError:
+        # File vanished between exists() and stat(); treat as absent.
+        return []
+
+    # Return copies so callers cannot mutate the cached entries.
+    return [dict(p) for p in _load_custom_providers_cached(str(CUSTOM_MODELS_FILE), mtime)]
+
+
+def clear_cache() -> None:
+    """Drop the memoized config.
+
+    The loader is keyed on the file's mtime, so a normal on-disk edit is
+    picked up automatically. This helper exists for the edge case of a
+    same-second rewrite (mtime unchanged) — notably in tests — and for
+    callers that want to force a re-read.
+    """
+    _load_custom_providers_cached.cache_clear()
 
 
 def get_custom_provider_choices() -> List[Tuple[str, str, str]]:
diff --git c/tradingagents/llm_clients/model_catalog.py w/tradingagents/llm_clients/model_catalog.py
index fccac7d..f1d16f0 100644
--- c/tradingagents/llm_clients/model_catalog.py
+++ w/tradingagents/llm_clients/model_catalog.py
@@ -212,6 +212,12 @@ def get_known_models() -> Dict[str, List[str]]:
     }
 
     for provider, mode_options in get_all_custom_model_options().items():
+        # Never let a custom provider shadow a built-in one. A custom entry
+        # keyed e.g. "openai" must not replace the curated OpenAI model list
+        # that validators.warn_if_unknown consumes (mirrors the
+        # existing_provider_keys guard in cli.utils._llm_provider_table).
+        if provider in MODEL_OPTIONS:
+            continue
         known_models[provider] = sorted(
             {
                 value
diff --git c/tradingagents/llm_clients/openai_client.py w/tradingagents/llm_clients/openai_client.py
index 0d78893..a653c1e 100644
--- c/tradingagents/llm_clients/openai_client.py
+++ w/tradingagents/llm_clients/openai_client.py
@@ -7,6 +7,7 @@ from langchain_openai import ChatOpenAI
 from .api_key_env import get_api_key_env
 from .base_client import BaseLLMClient, normalize_content
 from .capabilities import get_capabilities
+from .custom_provider_config import is_custom_openai_compatible_provider
 from .validators import validate_model
 
 
@@ -202,6 +203,31 @@ class OpenAIClient(BaseLLMClient):
         super().__init__(model, base_url, **kwargs)
         self.provider = provider.lower()
 
+    def _resolve_api_key(self, llm_kwargs: dict) -> None:
+        """Resolve and inject the provider's API key into ``llm_kwargs``.
+
+        ``get_api_key_env`` consults both the built-in provider map and any
+        user-defined custom provider config, so this works for built-in
+        providers (xai/deepseek/...) AND for custom OpenAI-compatible
+        providers declared in ``~/.tradingagents/custom_models.json`` with an
+        ``api_key_env``. When a provider has no configured env var (e.g.
+        Ollama, or a custom keyless gateway), fall back to the placeholder
+        ``"ollama"`` key the OpenAI SDK accepts for unauthenticated endpoints.
+        """
+        api_key_env = get_api_key_env(self.provider)
+        if api_key_env:
+            api_key = os.environ.get(api_key_env)
+            if api_key:
+                llm_kwargs["api_key"] = api_key
+            else:
+                raise ValueError(
+                    f"API key for provider '{self.provider}' is not set. "
+                    f"Please set the {api_key_env} environment variable "
+                    f"(e.g. add {api_key_env}=your_key to your .env file)."
+                )
+        else:
+            llm_kwargs["api_key"] = "ollama"
+
     def get_llm(self) -> Any:
         """Return configured ChatOpenAI instance."""
         self.warn_if_unknown_model()
@@ -212,21 +238,19 @@ class OpenAIClient(BaseLLMClient):
         # provider default so users can route through their own gateway.
         if self.provider in _PROVIDER_BASE_URL:
             llm_kwargs["base_url"] = self.base_url or _resolve_provider_base_url(self.provider)
-            api_key_env = get_api_key_env(self.provider)
-            if api_key_env:
-                api_key = os.environ.get(api_key_env)
-                if api_key:
-                    llm_kwargs["api_key"] = api_key
-                else:
-                    raise ValueError(
-                        f"API key for provider '{self.provider}' is not set. "
-                        f"Please set the {api_key_env} environment variable "
-                        f"(e.g. add {api_key_env}=your_key to your .env file)."
-                    )
-            else:
-                llm_kwargs["api_key"] = "ollama"
+            self._resolve_api_key(llm_kwargs)
         elif self.base_url:
             llm_kwargs["base_url"] = self.base_url
+            # Custom OpenAI-compatible providers (declared in
+            # custom_models.json) reach here: they carry a base_url but are
+            # not in the built-in _PROVIDER_BASE_URL map. Resolve their
+            # configured api_key_env too, otherwise auth would silently fall
+            # back to OPENAI_API_KEY and bill the wrong account. Built-in
+            # providers that land here (e.g. native openai with an explicit
+            # backend_url) keep relying on the SDK's implicit env pickup, so
+            # their existing behavior is unchanged.
+            if is_custom_openai_compatible_provider(self.provider):
+                self._resolve_api_key(llm_kwargs)
 
         # Forward user-provided kwargs
         for key in _PASSTHROUGH_KWARGS:
```

</details>

🤖 Automated patch via Claude Code — not pushed anywhere; yours to apply, adapt, or ignore.

**Comment by praxstack** [comment]
@AyushKar2005 I opened a PR **into your branch** with the fix (custom provider now reads its own `api_key_env` at runtime; +24 passing tests): AyushKar2005/TradingAgents#1 — merge, adapt, or cherry-pick as you like. 🤖

**Comment by Yijia-Xiao** [comment]
Thanks @AyushKar2005. Superseded by 20d3b07: any OpenAI-compatible endpoint works through the openai_compatible provider.
